src/smt/model_blocker.cpp

```diff
@@ -0,0 +1,288 @@
+          {
+            cond = cond.negate();
+            branch = cur[2];
+          }
+          impl = nm->mkNode(AND, cond, cpol ? branch : branch.negate());
+        }
+        else if ((catom.getKind() == EQUAL && catom[0].getType().isBoolean())
+                 || catom.getKind() == XOR)
+        {
+          // based on how the children evaluate in the model
+          std::vector<Node> children;
+          for (const Node& cn : catom)
+          {
+            Node vn = Node::fromExpr(m->getValue(cn.toExpr()));
+            Assert(vn.isConst());
+            children.push_back(vn.getConst<bool>() ? cn : cn.negate());
+          }
+          impl = nm->mkNode(AND, children);
+        }
+        else
+        {
+          // literals justified by themselves
+          visited[cur] = cur;
+          Trace("model-blocker-debug") << "...self justified" << std::endl;
+        }
+        if (visited[cur].isNull())
+        {
+          visit.push_back(cur);
+          if (impl.isNull())
+          {
+            Assert(cur.getKind() == AND);
+            Trace("model-blocker-debug") << "...recurse" << std::endl;
+            visit.insert(visit.end(), cur.begin(), cur.end());
+          }
+          else
+          {
+            Trace("model-blocker-debug")
+                << "...implicant : " << impl << std::endl;
+            implicant[cur] = impl;
+            visit.push_back(impl);
+          }
+        }
+      }
+      else if (it->second.isNull())
+      {
+        Node ret = cur;
+        it = implicant.find(cur);
+        if (it != implicant.end())
+        {
+          Node impl = it->second;
+          it = visited.find(impl);
+          Assert(it != visited.end());
+          Assert(!it->second.isNull());
+          ret = it->second;
+          Trace("model-blocker-debug")
+              << "...implicant res: " << ret << std::endl;
+        }
+        else
+        {
+          bool childChanged = false;
+          std::vector<Node> children;
+          // we never recurse to parameterized nodes
+          Assert(cur.getMetaKind() != metakind::PARAMETERIZED);
+          for (const Node& cn : cur)
+          {
+            it = visited.find(cn);
+            Assert(it != visited.end());
+            Assert(!it->second.isNull());
+            childChanged = childChanged || cn != it->second;
+            children.push_back(it->second);
+          }
+          if (childChanged)
+          {
+            ret = nm->mkNode(cur.getKind(), children);
+          }
+          Trace("model-blocker-debug") << "...recons res: " << ret << std::endl;
+        }
+        visited[cur] = ret;
+      }
+    } while (!visit.empty());
+    Assert(visited.find(formula) != visited.end());
+    Assert(!visited.find(formula)->second.isNull());
+    blocker = visited[formula].negate();
+  }
+  else
+  {
+    Assert(mode == BLOCK_MODELS_VALUES);
+    std::vector<Node> blockers;
+    // if specific terms were not specified, block all variables of
```

Comment:
Didn't you say that now we do not allow this case?

Reply:
No, we do not allow an empty set of values from the smt2 command: https://github.com/CVC4/CVC4/pull/3112/commits/8e6add556b0dae71218e79af9ecc486153677a00#diff-c78de56e9d21a5aa6d0deae68fa5c202R4421

We still allow it in the utility, as documented.

Reply:
But in the h file we have:
`We expect exprToBlock to be non-empty only if mode is BLOCK_MODELS_VALUES.`

Reply:
There is no `if`. :)